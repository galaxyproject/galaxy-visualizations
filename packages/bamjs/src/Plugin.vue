<script setup>
import { BamFile } from "@gmod/bam";
import axios from "axios";
import { computed, onMounted, ref, watch } from "vue";

const props = defineProps({
    root: String,
    tracks: Array,
    datasetId: String,
    datasetUrl: String,
    settings: {
        type: Object,
        default: () => ({
            max_records: 100,
            region_start: 0,
            region_end: 10000,
        }),
    },
    specs: Object,
});

const loading = ref(false);
const error = ref(null);
const bamData = ref(null);
const headerInfo = ref(null);
const records = ref([]);

const maxRecords = computed(() => props.settings?.max_records || 100);
const regionStart = computed(() => props.settings?.region_start || 0);
const regionEnd = computed(() => props.settings?.region_end || 10000);

async function loadBamFile() {
    loading.value = true;
    error.value = null;

    try {
        if (props.datasetUrl.startsWith("mock://")) {
            headerInfo.value = {
                version: "1.6",
                sortOrder: "coordinate",
                references: [
                    { name: "chr1", length: 248956422 },
                    { name: "chr2", length: 242193529 },
                    { name: "chrX", length: 156040895 },
                ],
                readGroups: [{ id: "HG001", sample: "NA12878", library: "lib_HG001", platform: "ILLUMINA" }],
                programs: [
                    {
                        id: "bwa",
                        name: "bwa",
                        version: "0.7.17",
                        commandLine: "bwa mem -M -t 16 GRCh38.fa input.fastq",
                    },
                ],
            };

            records.value = [
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:1234:2000",
                    refName: "chr1",
                    start: 69091,
                    end: 69141,
                    cigar: "50M",
                    seq: "ATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATC",
                    qual: "JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ",
                    flags: 99,
                    mapq: 60,
                    tags: { AS: 50, XS: 10, NM: 0, MD: "50" },
                },
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:1234:2000",
                    refName: "chr1",
                    start: 69141,
                    end: 69191,
                    cigar: "50M",
                    seq: "CGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT",
                    qual: "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIJ",
                    flags: 147,
                    mapq: 60,
                    tags: { AS: 50, XS: 8, NM: 0, MD: "50" },
                },
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:5678:2500",
                    refName: "chr2",
                    start: 158312,
                    end: 158362,
                    cigar: "25M1I24M",
                    seq: "GACTGCAGACTGCAGACTGCAGAACTGCAGACTGCAGACTGCAGACTGCAG",
                    qual: "HHHHHHHHHHHHHHHHHHHHHHHHHGGGGGGGGGGGGGGGGGGGGGGGGGGG",
                    flags: 0,
                    mapq: 42,
                    tags: { AS: 48, XS: 5, NM: 1, MD: "25^A24" },
                },
                {
                    name: "HWI-D00119:50:H7AP8ADXX:1:1101:9012:3000",
                    refName: "chr1",
                    start: 52381,
                    end: 52426,
                    cigar: "45M5S",
                    seq: "TGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGC",
                    qual: "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF?????",
                    flags: 16,
                    mapq: 23,
                    tags: { AS: 40, XS: 15, NM: 2, MD: "20A10T14" },
                },
            ];

            return;
        }

        if (props.datasetUrl.startsWith("http")) {
            let bamFile;

            try {
                bamFile = new BamFile({
                    bamUrl: props.datasetUrl,
                    baiUrl: props.datasetUrl + ".bai",
                });

                headerInfo.value = await bamFile.getHeader();
            } catch (indexError) {
                console.warn("Index file not found, trying without index:", indexError.message);

                bamFile = new BamFile({
                    bamUrl: props.datasetUrl,
                });

                headerInfo.value = await bamFile.getHeader();
            }

            const refSeqs = headerInfo.value.references;
            if (refSeqs && refSeqs.length > 0) {
                const firstRef = refSeqs[0];
                const endPos = Math.min(regionEnd.value, firstRef.length);

                try {
                    const bamRecords = await bamFile.getRecordsForRange(firstRef.name, regionStart.value, endPos, {
                        maxRecords: maxRecords.value,
                    });

                    records.value = bamRecords;
                } catch (rangeError) {
                    console.warn("Could not get records for range:", rangeError.message);
                    records.value = [];
                }
            }
        } else {
            const response = await axios.get(props.datasetUrl);

            if (response.data.download_url) {
                const bamFile = new BamFile({
                    bamUrl: response.data.download_url,
                });

                headerInfo.value = await bamFile.getHeader();

                const refSeqs = headerInfo.value.references;
                if (refSeqs && refSeqs.length > 0) {
                    const firstRef = refSeqs[0];
                    const endPos = Math.min(regionEnd.value, firstRef.length);

                    const bamRecords = await bamFile.getRecordsForRange(firstRef.name, regionStart.value, endPos, {
                        maxRecords: maxRecords.value,
                    });

                    records.value = bamRecords;
                }
            } else {
                error.value = "Unable to access BAM file data";
            }
        }
    } catch (err) {
        error.value = `Error loading BAM file: ${err.message}`;
        console.error("BAM loading error:", err);
    } finally {
        loading.value = false;
    }
}

function formatRecord(record) {
    const flags = [];
    if (record.flags & 0x1) flags.push("paired");
    if (record.flags & 0x2) flags.push("proper_pair");
    if (record.flags & 0x4) flags.push("unmapped");
    if (record.flags & 0x8) flags.push("mate_unmapped");
    if (record.flags & 0x10) flags.push("reverse");
    if (record.flags & 0x20) flags.push("mate_reverse");
    if (record.flags & 0x40) flags.push("read1");
    if (record.flags & 0x80) flags.push("read2");
    if (record.flags & 0x100) flags.push("secondary");
    if (record.flags & 0x200) flags.push("qc_fail");
    if (record.flags & 0x400) flags.push("duplicate");
    if (record.flags & 0x800) flags.push("supplementary");

    return {
        name: record.name || "unknown",
        refName: record.refName || "unknown",
        start: record.start,
        end: record.end,
        cigar: record.cigar || "",
        seq: record.seq || "",
        qual: record.qual || "",
        flags: flags.join(","),
        mapq: record.mapq || 0,
        tags: record.tags
            ? Object.entries(record.tags)
                  .map(([k, v]) => `${k}:${v}`)
                  .join(";")
            : "",
    };
}

function formatHeader(header) {
    let output = "BAM Header Information:\n";
    output += "======================\n\n";

    if (header.version) {
        output += `Version: ${header.version}\n`;
    }

    if (header.sortOrder) {
        output += `Sort Order: ${header.sortOrder}\n`;
    }

    if (header.groupOrder) {
        output += `Group Order: ${header.groupOrder}\n`;
    }

    output += "\nReference Sequences:\n";
    output += "-------------------\n";
    if (header.references) {
        header.references.forEach((ref, index) => {
            output += `${index + 1}. ${ref.name} (length: ${ref.length})\n`;
        });
    }

    if (header.readGroups && header.readGroups.length > 0) {
        output += "\nRead Groups:\n";
        output += "------------\n";
        header.readGroups.forEach((rg, index) => {
            output += `${index + 1}. ID: ${rg.id}\n`;
            if (rg.sample) output += `   Sample: ${rg.sample}\n`;
            if (rg.library) output += `   Library: ${rg.library}\n`;
            if (rg.platform) output += `   Platform: ${rg.platform}\n`;
        });
    }

    if (header.programs && header.programs.length > 0) {
        output += "\nPrograms:\n";
        output += "---------\n";
        header.programs.forEach((prog, index) => {
            output += `${index + 1}. ID: ${prog.id}\n`;
            if (prog.name) output += `   Name: ${prog.name}\n`;
            if (prog.version) output += `   Version: ${prog.version}\n`;
            if (prog.commandLine) output += `   Command: ${prog.commandLine}\n`;
        });
    }

    return output;
}

function formatRecords(recordList) {
    if (!recordList || recordList.length === 0) {
        return "No records found in the specified region.";
    }

    let output = `\nBAM Records (showing ${recordList.length} records):\n`;
    output += "=".repeat(50) + "\n\n";

    recordList.forEach((record, index) => {
        const formatted = formatRecord(record);
        output += `Record ${index + 1}:\n`;
        output += `  Name: ${formatted.name}\n`;
        output += `  Reference: ${formatted.refName}\n`;
        output += `  Position: ${formatted.start}-${formatted.end}\n`;
        output += `  CIGAR: ${formatted.cigar}\n`;
        output += `  MAPQ: ${formatted.mapq}\n`;
        output += `  Flags: ${formatted.flags}\n`;
        if (formatted.seq) {
            output += `  Sequence: ${formatted.seq.substring(0, 100)}${formatted.seq.length > 100 ? "..." : ""}\n`;
        }
        if (formatted.qual) {
            output += `  Quality: ${formatted.qual.substring(0, 100)}${formatted.qual.length > 100 ? "..." : ""}\n`;
        }
        if (formatted.tags) {
            output += `  Tags: ${formatted.tags}\n`;
        }
        output += "\n";
    });

    return output;
}

watch(() => props.datasetUrl, loadBamFile);

onMounted(() => {
    if (props.datasetUrl) {
        loadBamFile();
    }
});
</script>

<template>
    <div class="bam-viewer">
        <div v-if="loading" class="loading">Loading BAM file...</div>

        <div v-else-if="error" class="error">
            {{ error }}
        </div>

        <div v-else-if="headerInfo || records.length > 0" class="bam-records">
            <div v-if="headerInfo">{{ formatHeader(headerInfo) }}</div>
            <div v-if="records.length > 0">{{ formatRecords(records) }}</div>
        </div>

        <div v-else class="loading">No BAM data available</div>
    </div>
</template>
