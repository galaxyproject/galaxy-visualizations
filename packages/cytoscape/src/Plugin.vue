<script setup lang="ts">
import Cytoscape, { type Core, type ElementDefinition, type ElementsDefinition } from "cytoscape";
import { onMounted, onUnmounted, ref, watch } from "vue";

import { parseSIF, runSearchAlgorithm, runTraversalType, styleGenerator } from "./utils";

const TEST_URL = "http://cdn.jsdelivr.net/gh/galaxyproject/galaxy-test-data/1.cytoscapejson";

type Settings = {
    directed: boolean;
    layout_name: "breadthfirst" | "circle" | "concentric" | "cose" | "grid" | "null" | "preset" | "random";
    curve_style: "haystack" | "straight" | "bezier" | "unbundled-bezier" | "segments" | "taxi";
    color_picker_nodes: string;
    color_picker_edges: string;
    color_picker_highlighted: string;
    graph_traversal?: "successors" | "predecessors" | "outgoers" | "incomers" | "roots" | "leaves";
    search_algorithm?: "bfs" | "dfs" | "kruskal" | "astar";
};

interface Props {
    root: string;
    datasetId: string;
    datasetUrl: string;
    settings: Settings;
}

const props = defineProps<Props>();

const viewport = ref(null);

const cytoscape = ref<Core | null>(null);
const dataset = ref<ElementsDefinition | ElementDefinition[]>();

async function render() {
    cytoscape.value = Cytoscape({
        container: viewport.value,
        elements: dataset.value,
        layout: {
            name: props.settings.layout_name,
            idealEdgeLength: () => 100,
            nodeOverlap: 20,
        },
        minZoom: 0.1,
        maxZoom: 50,
        style: styleGenerator(props.settings),
    });

    if (props.settings.search_algorithm === "kruskal") {
        //@ts-ignore
        cytoscape.value.elements().kruskal().edges().addClass("searchpath");
    }

    cytoscape.value.on("tap", "node", (e) => {
        const ele = e.target;

        const search_algorithm = props.settings.search_algorithm;
        const traversal_type = props.settings.graph_traversal;

        if (cytoscape.value && search_algorithm) {
            runSearchAlgorithm(cytoscape.value, ele.id(), search_algorithm);
        } else if (cytoscape.value && traversal_type) {
            runTraversalType(cytoscape.value, ele.id(), traversal_type);
        }
    });
}

async function getDataset() {
    try {
        const datasetUrl = props.datasetId === "__test__" ? TEST_URL : props.datasetUrl;
        const response = await fetch(datasetUrl);
        if (response.ok) {
            const content = await response.json();
            if (content.file_ext === "sif") {
                dataset.value = parseSIF(content).content;
            } else {
                dataset.value = content.elements ?? content;
            }
        } else {
            console.error("Fetch failed", response);
        }
    } catch (err) {
        console.error("Dataset fetch error", err);
    }
}

watch(
    () => props,
    () => render(),
    { deep: true },
);

onMounted(async () => {
    await getDataset();
    await render();
    window.addEventListener("resize", function () {
        cytoscape.value?.resize();
    });
});

onUnmounted(() => {
    window.removeEventListener("resize", function () {
        console.log("Removing resize event listener");
    });

    cytoscape.value?.destroy();
});
</script>

<template>
    <div class="h-screen overflow-hidden" ref="viewport"></div>
</template>
