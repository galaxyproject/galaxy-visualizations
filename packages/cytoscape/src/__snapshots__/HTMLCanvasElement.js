HTMLCanvasElement.prototype.toDataURL = vi.fn(() => "data:image/png;base64,MOCKDATA");

HTMLCanvasElement.prototype.getContext = vi.fn(function () {
    const canvas = this;
    if (!canvas._opsContainer) {
        const opsContainer = document.createElement("div");
        opsContainer.setAttribute("data-canvas-ops", "");
        canvas.insertAdjacentElement("afterend", opsContainer);
        canvas._opsContainer = opsContainer;
    }
    const opsContainer = canvas._opsContainer;
    function logOp(name, ...args) {
        const div = document.createElement("div");
        div.textContent = `${name}(${args.map((a) => JSON.stringify(a)).join(", ")})`;
        opsContainer.appendChild(div);
    }
    return {
        fillRect(x, y, w, h) {
            logOp("fillRect", x, y, w, h);
        },
        clearRect(x, y, w, h) {
            logOp("clearRect", x, y, w, h);
        },
        getImageData(x, y, w, h) {
            logOp("getImageData", x, y, w, h);
            return { data: new Uint8ClampedArray(w * h * 4) };
        },
        putImageData(imageData, x, y) {
            logOp("putImageData", x, y);
        },
        createImageData(w, h) {
            logOp("createImageData", w, h);
            return { data: new Uint8ClampedArray(w * h * 4) };
        },
        setTransform() {
            logOp("setTransform");
        },
        drawImage() {
            logOp("drawImage");
        },
        save() {
            logOp("save");
        },
        fillText(text, x, y) {
            logOp("fillText", text, x, y);
        },
        restore() {
            logOp("restore");
        },
        measureText() {
            logOp("measureText");
            return { width: 42 };
        },
        translate(x, y) {
            logOp("translate", x, y);
        },
        beginPath() {
            logOp("beginPath");
        },
        arc(x, y, r) {
            logOp("arc", x, y, r);
        },
        fill() {
            logOp("fill");
        },
        closePath() {
            logOp("closePath");
        },
        moveTo(x, y) {
            logOp("moveTo", x, y);
        },
        lineTo(x, y) {
            logOp("lineTo", x, y);
        },
        rotate(a) {
            logOp("rotate", a);
        },
        scale(x, y) {
            logOp("scale", x, y);
        },
        stroke() {
            logOp("stroke");
        },
        strokeRect(x, y, w, h) {
            logOp("strokeRect", x, y, w, h);
        },
        clip() {
            logOp("clip");
        },
    };
});
