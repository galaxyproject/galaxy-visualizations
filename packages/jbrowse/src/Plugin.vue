<script setup>
import { onMounted, ref, watch } from "vue";
import { renderJBrowse } from "./render";

const props = defineProps({
    datasetId: String,
    datasetUrl: String,
    root: String,
    settings: Object,
    specs: Object,
    tracks: Array,
});

const viewport = ref(null);

function render() {
    const config = JSON.parse(JSON.stringify(props.settings.jbrowseConfig));
    if (viewport.value) {
        renderJBrowse(viewport.value, config);
    }
}

onMounted(() => {
    render();
});

watch(
    () => props,
    () => render(),
    { deep: true },
);
</script>

<template>
    <div ref="viewport"></div>
</template>
