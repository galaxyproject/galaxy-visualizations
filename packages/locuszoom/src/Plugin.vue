<script setup>
import { onMounted, watch, ref } from "vue";
import "locuszoom/dist/locuszoom.css";
import LocusZoom from "locuszoom/esm";
import LzTabixSource from "locuszoom/esm/ext/lz-tabix-source";
import { makeGWASParser } from "locuszoom/esm/ext/lz-parsers";
import { plotUpdatesUrl, plotWatchesUrl } from "locuszoom/esm/ext/lz-dynamic-urls";

LocusZoom.use(LzTabixSource);

window.nowrap = false;
window.mode = false;

const COLOR = "#E30A17";
const MAX_RANGE = 10000000;
const SIZE = 40;

const props = defineProps({
    datasetId: String,
    datasetUrl: String,
    root: String,
    settings: Object,
    specs: Object,
    tracks: Array,
});

const errorMessage = ref("");

function render() {
    const id = props.settings.tabix?.id;
    const chrIn = props.settings.chromosome;
    const startIn = props.settings.start;
    const endIn = props.settings.end;
    const chrom_col = props.settings.chrom_col;
    const pos_col = props.settings.pos_col;
    const ref_col = props.settings.ref_col;
    const alt_col = props.settings.alt_col;
    const pvalue_col = props.settings.pvalue_col;
    const is_neg_log_pvalue = props.settings.is_neg_log_pvalue;
    const beta_col = props.settings.beta_col;
    const stderr_beta_col = props.settings.stderr_beta_col;
    if (!id) {
        errorMessage.value = "Please select a Tabix file.";
        return;
    }
    if (endIn - startIn > MAX_RANGE) {
        errorMessage.value = "We cannot output more than 10Mb at a time!";
        return;
    }
    if (endIn <= startIn) {
        errorMessage.value = "Invalid input: End position must be bigger than start position!";
        return;
    }
    errorMessage.value = "";
    const gwasParser = makeGWASParser({
        chrom_col,
        pos_col,
        ref_col,
        alt_col,
        pvalue_col,
        is_neg_log_pvalue,
        beta_col,
        stderr_beta_col,
    });
    let data_sources = new LocusZoom.DataSources().add("assoc", [
        "TabixUrlSource",
        {
            url_data: `${props.root}api/datasets/${props.datasetId}/display`,
            url_tbi: `${props.root}api/datasets/${id}/display`,
            parser_func: gwasParser,
            overfetch: 0,
        },
    ]);
    let stateUrlMapping = { chr: "chrom", start: "start", end: "end", ldrefvar: "ld_variant" };
    let association_data_layer_mods = {
        id: "associationpvalues_",
        name: "association",
        point_shape: "circle",
        point_size: SIZE,
        color: COLOR,
        legend: [{ shape: "circle", color: COLOR, size: SIZE, label: "phen", class: "lz-data_layer-scatter" }],
        tooltip: {
            closable: true,
            show: { or: ["highlighted", "selected"] },
            hide: { and: ["unhighlighted", "unselected"] },
            html: `<strong>phen</strong><br><strong>{{assoc:variant|htmlescape}}</strong><br>
                P Value: <strong>{{assoc:log_pvalue|logtoscinotation|htmlescape}}</strong><br>
                Ref. Allele: <strong>{{assoc:ref_allele|htmlescape}}</strong><br>`,
        },
    };
    const data_layer = {
        ...LocusZoom.Layouts.get("data_layer", "association_pvalues", association_data_layer_mods),
        namespace: { assoc: "assoc" },
        data_operations: [],
    };
    const panel = LocusZoom.Layouts.get("panel", "association", {
        data_layers: [data_layer],
    });
    let plot = LocusZoom.populate("#lz-plot", data_sources, {
        state: { chr: chrIn, start: startIn, end: endIn },
        responsive_resize: true,
        panels: [panel],
        toolbar: LocusZoom.Layouts.get("toolbar", "standard_plot"),
    });
    plotUpdatesUrl(plot, stateUrlMapping);
    plotWatchesUrl(plot, stateUrlMapping);

    // manually set size of svg container
    document.querySelector("#lz-plot_svg").style.height = "calc(100vh - 50px)";
}

onMounted(() => {
    render();
});

watch(
    () => props,
    () => {
        render();
    },
    { deep: true },
);
</script>

<template>
    <div id="lz-plot" class="h-100"></div>
    <div v-if="errorMessage" class="mt-4 p-3 border bg-red-100 text-red-700 text-sm text-center rounded">
        {{ errorMessage }}
    </div>
</template>

<style>
.lz-panel-toolbar {
    display: none;
}
</style>
