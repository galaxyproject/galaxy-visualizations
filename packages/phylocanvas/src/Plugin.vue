<script setup>
import { ArrowPathIcon, ArrowLeftIcon, ArrowRightIcon } from "@heroicons/vue/24/outline";
import { NAlert, NButton, NIcon, NInputNumber, NTooltip } from "naive-ui";
import { onBeforeUnmount, onMounted, ref, watch } from "vue";

import phylocanvas from "@/phylocanvas.min.js";

const TEST_URL = "http://cdn.jsdelivr.net/gh/galaxyproject/galaxy-test-data/1.newick";

const props = defineProps({
    datasetId: String,
    datasetUrl: String,
    settings: Object,
});

let tree = null;

const container = ref(null);
const errorMessage = ref("");
const isLoading = ref(true);
const treeindex = ref(0);
const treeindexInput = ref(1);
const treelist = ref([]);
const viewport = ref(null);

function extract(content) {
    treelist.value = [];
    let depth = 0,
        start = 0;
    for (let i = 0; i < content.length; i++) {
        const char = content[i];
        if (char === "(") {
            depth++;
        } else if (char === ")") {
            depth--;
        } else if (char === ";" && depth === 0) {
            treelist.value.push(content.slice(start, i + 1).trim());
            start = i + 1;
        }
    }
    if (treelist.value.length === 0) {
        errorMessage.value = "No valid trees found.";
    }
}

function handleResize() {
    if (tree && viewport.value) {
        const bcr = viewport.value.getBoundingClientRect();
        tree.resize(bcr.width, bcr.height);
    }
}

async function load() {
    isLoading.value = true;
    const datasetUrl = props.datasetId === "__test__" ? TEST_URL : props.datasetUrl;
    const response = await fetch(datasetUrl);
    if (response.ok) {
        const content = await response.text();
        extract(content);
        render(0);
    } else {
        errorMessage.value = "Failed to load data.";
        console.error(response);
    }
    isLoading.value = false;
}

function onInputIndex(val) {
    if (val >= 1 && val <= treelist.value.length) {
        render(val - 1);
    }
}

function render(newIndex) {
    if (newIndex >= 0 && newIndex < treelist.value.length) {
        treeindex.value = newIndex;
        const source = treelist.value[newIndex];
        if (tree) {
            tree.destroy();
            tree = null;
        }
        try {
            tree = new phylocanvas(container.value, {
                source,
                alignLabels: props.settings.align_labels,
                fillColour: props.settings.node_color,
                highlightColour: props.settings.highlighted_color,
                interactive: true,
                nodeShape: props.settings.node_shape,
                showLabels: props.settings.show_labels,
                showLeafLabels: props.settings.show_labels,
                strokeColour: props.settings.edge_color,
                type: props.settings.tree_type,
            });
        } catch (e) {
            errorMessage.value = "Failed to render tree.";
            console.error(errorMessage.value, e);
        }
    } else {
        errorMessage.value = `Invalid tree index: ${newIndex}`;
        console.error(errorMessage.value);
    }
}

onMounted(() => {
    load();
    window.addEventListener("resize", handleResize);
});

onBeforeUnmount(() => {
    window.removeEventListener("resize", handleResize);
});

watch(
    () => props.settings,
    () => render(treeindex.value),
    { deep: true },
);

watch(treeindex, (newVal) => {
    treeindexInput.value = newVal + 1;
});
</script>

<template>
    <n-alert v-if="errorMessage" type="error" class="m-1">
        {{ errorMessage }}
    </n-alert>
    <div v-if="isLoading" class="m-2">
        <ArrowPathIcon class="animate-spin size-4 inline mx-1" />
        <span class="text-xs">Please wait...</span>
    </div>
    <div ref="viewport" class="h-screen overflow-hidden">
        <div ref="container" class="h-screen"></div>
        <div
            v-if="treelist.length > 1"
            class="absolute bottom-2 left-1/2 transform -translate-x-1/2 bg-white opacity-[0.9] rounded px-3 py-2">
            <div class="flex items-center space-x-2">
                <n-tooltip trigger="hover" :to="false">
                    <template #trigger>
                        <n-button size="small" :disabled="treeindex === 0" @click="render(treeindex - 1)">
                            <template #icon>
                                <n-icon><ArrowLeftIcon /></n-icon>
                            </template>
                        </n-button>
                    </template>
                    <span class="text-xs">Previous</span>
                </n-tooltip>
                <n-input-number
                    v-model:value="treeindexInput"
                    :min="1"
                    :max="treelist.length"
                    :show-button="false"
                    placeholder="?"
                    size="small"
                    style="width: 4rem"
                    @update:value="onInputIndex" />
                <span class="text-sm">of {{ treelist.length }}</span>
                <n-tooltip trigger="hover" :to="false">
                    <template #trigger>
                        <n-button
                            size="small"
                            :disabled="treeindex === treelist.length - 1"
                            @click="render(treeindex + 1)">
                            <template #icon>
                                <n-icon><ArrowRightIcon /></n-icon>
                            </template>
                        </n-button>
                    </template>
                    <span class="text-xs">Next</span>
                </n-tooltip>
            </div>
        </div>
    </div>
</template>
