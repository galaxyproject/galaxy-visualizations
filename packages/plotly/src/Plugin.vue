<script setup>
import Plotly from "plotly.js-dist";
import { onMounted, ref, watch } from "vue";

import plotlySurface from "@/variants/surface";
import plotlyBasics from "@/variants/basics";
import plotlyBox from "@/variants/box";
import plotlyHeatmap from "@/variants/heatmap";
import plotlyHistogram from "@/variants/histogram";

const props = defineProps({
    datasetId: String,
    datasetUrl: String,
    root: String,
    settings: Object,
    specs: Object,
    tracks: Array,
});

const message = ref();
const viewport = ref(null);

async function render() {
    let wrapper = null;
    switch (props.specs?.variant) {
        case "box":
            wrapper = plotlyBox;
            break;
        case "histogram":
            wrapper = plotlyHistogram;
            break;
        case "heatmap":
            wrapper = plotlyHeatmap;
            break;
        case "surface":
            wrapper = plotlySurface;
            break;
        default:
            wrapper = plotlyBasics;
    }
    try {
        message.value = "";
        const { data, layout, config } = await wrapper(props.datasetId, props.settings, props.tracks);
        Plotly.newPlot(viewport.value, data, layout, config);
    } catch (e) {
        message.value = `Failed to render: ${e}`;
        console.error(e);
    }
}

onMounted(() => {
    render();
});

watch(
    () => props,
    () => render(),
    { deep: true },
);
</script>

<template>
    <div v-if="message" class="bg-sky-100 border border-sky-200 p-1 rounded text-sky-800 text-sm">
        {{ message }}
    </div>
    <div ref="viewport" class="h-screen" />
</template>
