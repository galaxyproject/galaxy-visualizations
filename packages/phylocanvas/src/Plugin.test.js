import { mount } from "@vue/test-utils";
import { describe, expect, it, vi } from "vitest";

import Plugin from "@/Plugin.vue";

import "@/__snapshots__/HTMLCanvasElement.js";

const content =
    "((raccoon:19.19959,bear:6.80041):0.84600,((sea_lion:11.99700, seal:12.00300):7.52973,((monkey:100.85930,cat:47.14069):20.59201, weasel:18.87953):2.09460):3.87382,dog:25.46154);";

vi.spyOn(global, "fetch").mockResolvedValue({
    ok: true,
    text: async () => content,
});

describe("Plugin", () => {
    it("should render correctly", () => {
        const wrapper = mount(Plugin, {
            props: {
                datasetId: "MY_DATASET_ID",
                datasetUrl: "phylocanvas.nwk",
                root: "/",
                settings: {},
                tracks: [],
            },
        });
        const html = wrapper.html().replace(/__file=".*\/packages/g, '__file="<root>/packages');
        expect(html).toMatchSnapshot();
    });
});
