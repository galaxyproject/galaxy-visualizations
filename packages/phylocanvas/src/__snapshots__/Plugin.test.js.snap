// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Plugin > should render correctly 1`] = `
"<transition-stub name="fade-in-height-expand-transition" appear="false" persisted="false" css="true">
  <div class="m-1 n-alert n-alert--show-icon" style="--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(251, 238, 241, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #d03050; --n-border: 1px solid rgba(243, 203, 211, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;" role="alert">
    <!---->
    <div class="n-alert__border"></div>
    <div class="n-alert__icon" aria-hidden="true"><i class="n-base-icon"><svg viewBox="0 0 48 48" version="1.1" xmlns="http://www.w3.org/2000/svg">
          <g stroke="none" stroke-width="1" fill-rule="evenodd">
            <g fill-rule="nonzero">
              <path d="M24,4 C35.045695,4 44,12.954305 44,24 C44,35.045695 35.045695,44 24,44 C12.954305,44 4,35.045695 4,24 C4,12.954305 12.954305,4 24,4 Z M17.8838835,16.1161165 L17.7823881,16.0249942 C17.3266086,15.6583353 16.6733914,15.6583353 16.2176119,16.0249942 L16.1161165,16.1161165 L16.0249942,16.2176119 C15.6583353,16.6733914 15.6583353,17.3266086 16.0249942,17.7823881 L16.1161165,17.8838835 L22.233,24 L16.1161165,30.1161165 L16.0249942,30.2176119 C15.6583353,30.6733914 15.6583353,31.3266086 16.0249942,31.7823881 L16.1161165,31.8838835 L16.2176119,31.9750058 C16.6733914,32.3416647 17.3266086,32.3416647 17.7823881,31.9750058 L17.8838835,31.8838835 L24,25.767 L30.1161165,31.8838835 L30.2176119,31.9750058 C30.6733914,32.3416647 31.3266086,32.3416647 31.7823881,31.9750058 L31.8838835,31.8838835 L31.9750058,31.7823881 C32.3416647,31.3266086 32.3416647,30.6733914 31.9750058,30.2176119 L31.8838835,30.1161165 L25.767,24 L31.8838835,17.8838835 L31.9750058,17.7823881 C32.3416647,17.3266086 32.3416647,16.6733914 31.9750058,16.2176119 L31.8838835,16.1161165 L31.7823881,16.0249942 C31.3266086,15.6583353 30.6733914,15.6583353 30.2176119,16.0249942 L30.1161165,16.1161165 L24,22.233 L17.8838835,16.1161165 L17.7823881,16.0249942 L17.8838835,16.1161165 Z"></path>
            </g>
          </g>
        </svg></i></div>
    <div class="n-alert-body n-alert-body--bordered">
      <!---->
      <div class="n-alert-body__content">Failed to render tree.</div>
    </div>
  </div>
</transition-stub>
<!--v-if-->
<div class="h-screen overflow-hidden">
  <div class="h-screen"></div>
  <!--v-if-->
</div>"
`;
